### PR TITLE
Gorm upgrade

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

# TL;DR
Upgrading GORM to V2
Removing unnecessary fields and fixing foreign key annotations for gorm V2.

Upgrading to V2, has converted to ID any other integer column from int to bigint

Testing done : 
Launching subworkflow and greet examples from cookbook and using flytectl to get execution details.

Comparing the schemas:

eg : 

```
 sdiff -s exec_events_old exec_events_new
 id                | integer                  |           | n |	 id                | bigint                   |           | n



 sdiff -s exec_old exec_new
 id                       | integer                  |        |	 id                       | bigint                   |       
 launch_plan_id           | integer                  |        |	 launch_plan_id           | bigint                   |       
 workflow_id              | integer                  |        |	 workflow_id              | bigint                   |       
 task_id                  | integer                  |        |	 task_id                  | bigint                   |       
 source_execution_id      | integer                  |        |	 source_execution_id      | bigint                   |       
 parent_node_execution_id | integer                  |        |	 parent_node_execution_id | bigint                   |  

```

## Type
 - [X] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue

fixes https://github.com/flyteorg/flyte/issues/1043

## Follow-up issue
https://github.com/flyteorg/flyte/issues/1955
https://github.com/flyteorg/flyte/issues/1956

### CODE REVIEW

File: cmd/entrypoints/migrate.go [hunk 1]
```diff
@@ -44,15 +45,18 @@ var migrateCmd = &cobra.Command{
...
 		if err != nil {
 			logger.Fatal(ctx, err)
 		}
-		defer db.Close()
-		db.LogMode(true)
-		if err = db.DB().Ping(); err != nil {
-			logger.Fatal(ctx, err)
-		}
+		//defer db.Close()
+		//db.LogMode(true)
```
Comment (katrogan) [inline cmd/entrypoints/migrate.go:56]: why is the call to `LogMode(true)` commented out? it should work in the latest gorm too

File: pkg/repositories/errors/postgres.go [hunk 2]
```diff
@@ -13,7 +13,7 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 
 	"github.com/flyteorg/flyteadmin/pkg/errors"
-	"github.com/jinzhu/gorm"
+	"gorm.io/gorm"
```
Comment (katrogan) [inline pkg/repositories/errors/postgres.go:16]: can you add the changes from https://github.com/flyteorg/datacatalog/commit/11d64b3479a33a60a77a3232dad2b13489ebef5a#diff-7ea834bee5dcbc7c2b02c416df4ecb4e5d85f4776311e6d114b64140cbcb1251R43 to this file too? 

we ran into issues after the gorm upgrade that necessitated adding them in

Comment (pmahindrakar-oss) [reply pkg/repositories/errors/postgres.go]: Done

File: cmd/entrypoints/migrate.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply cmd/entrypoints/migrate.go]: Fixed it.

**Comment by katrogan** [comment]
hey @pmahindrakar-oss can you verify that you can run executions end to end with this image?

**Comment by pmahindrakar-oss** [comment]
@katrogan @EngHabu  PTAL

File: pkg/repositories/models/execution.go [hunk 3]
```diff
@@ -34,16 +34,13 @@ type Execution struct {
 	// Prefixed with Execution to avoid clashes with gorm.Model UpdatedAt
 	ExecutionUpdatedAt *time.Time
 	Duration           time.Duration
-	ExecutionEvents    []ExecutionEvent
```
Comment (katrogan) [inline pkg/repositories/models/execution.go:37]: why is this deleted? how do we store the relationship now?

File: pkg/repositories/models/execution.go [hunk 4]
```diff
@@ -34,16 +34,13 @@ type Execution struct {
...
 	// In the case of an aborted execution this string may be non-empty.
 	// It should be ignored for any other value of phase other than aborted.
 	AbortCause string `valid:"length(0|255)"`
 	// Corresponds to the execution mode used to trigger this execution
 	Mode int32
 	// The "parent" execution (if there is one) that is related to this execution.
 	SourceExecutionID uint
-	// Descendant execution related to this execution.
-	DescendantExecution *Execution
```
Comment (katrogan) [inline pkg/repositories/models/execution.go:46]: ditto

File: tests/bootstrap.go [hunk 5]
```diff
@@ -46,31 +47,57 @@ func truncateAllTablesForTestingOnly() {
...
 	TruncateLaunchPlans := fmt.Sprintf("TRUNCATE TABLE launch_plans;")
-	// HACK: alter executions table so that spec is not required for testing.
-	TruncateExecutions := fmt.Sprintf("TRUNCATE TABLE executions; alter table executions alter column spec drop not null;")
+
 	TruncateExecutionEvents := fmt.Sprintf("TRUNCATE TABLE execution_events;")
 	TruncateNamedEntityMetadata := fmt.Sprintf("TRUNCATE TABLE named_entity_metadata;")
-	TruncateNodeExecutions := fmt.Sprintf("TRUNCATE TABLE node_executions;")
 	TruncateNodeExecutionEvents := fmt.Sprintf("TRUNCATE TABLE node_execution_events;")
-	TruncateTaskExecutions := fmt.Sprintf("TRUNCATE TABLE task_executions;")
+	TruncateTaskNodeExecutions := fmt.Sprintf("truncate table task_executions,node_executions,executions;")
```
Comment (katrogan) [inline tests/bootstrap.go:54]: is this change because of foreign key references? can you add a comment explaining why it's necessary?

File: tests/bootstrap.go [hunk 6]
```diff
@@ -46,31 +47,57 @@ func truncateAllTablesForTestingOnly() {
...
 	db.Exec(TruncateTasks)
 	db.Exec(TruncateWorkflows)
 	db.Exec(TruncateLaunchPlans)
-	db.Exec(TruncateExecutions)
 	db.Exec(TruncateExecutionEvents)
 	db.Exec(TruncateNamedEntityMetadata)
-	db.Exec(TruncateNodeExecutions)
 	db.Exec(TruncateNodeExecutionEvents)
-	db.Exec(TruncateTaskExecutions)
+	db.Exec(TruncateTaskNodeExecutions)
```
Comment (katrogan) [inline tests/bootstrap.go:79]:
```suggestion
	db.Exec(TruncateAllExecutions)
```

**Review by katrogan — COMMENTED** [review]
hey @pmahindrakar-oss thanks so much for this massive undertaking!

are you able to stand up this admin instance on a sandbox and verify that you can launch executions and use the console to navigate esp. with dynamic tasks, child workflows and other more complicated cases?

File: pkg/repositories/gormimpl/execution_repo.go [hunk 7]
```diff
@@ -43,15 +43,15 @@ func (r *ExecutionRepo) Get(ctx context.Context, input interfaces.Identifier) (m
 		},
 	}).Take(&execution)
 	timer.Stop()
-	if tx.Error != nil {
-		return models.Execution{}, r.errorTransformer.ToFlyteAdminError(tx.Error)
-	}
-	if tx.RecordNotFound() {
-		return models.Execution{}, errors.GetMissingEntityError("execution", &core.Identifier{
+
+	if tx.Error != nil && errors.Is(tx.Error, gorm.ErrRecordNotFound) {
```
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/repositories/gormimpl/execution_repo.go:47]: actually, doesn't necessarily need to be in this PR - we can file a housekeeping issue to track

File: pkg/repositories/models/execution.go (hunk 3 above)
Comment (pmahindrakar-oss) [reply pkg/repositories/models/execution.go:37]: I don't find this field is being read anywhere . Please let me know how we are using it .

File: pkg/repositories/models/execution.go (hunk 4 above)
Comment (pmahindrakar-oss) [reply pkg/repositories/models/execution.go:46]: Same here

File: tests/bootstrap.go (hunk 5 above)
Comment (pmahindrakar-oss) [reply tests/bootstrap.go]: I have removed this change now as we are disabling foreign key constraints during migration which prevents gorm to add these constraints to the tables.

File: tests/bootstrap.go (hunk 6 above)
Comment (pmahindrakar-oss) [reply tests/bootstrap.go]: reverted this

File: pkg/repositories/config/migrations.go [hunk 8]
```diff
@@ -124,7 +124,7 @@ var Migrations = []*gormigrate.Migration{
 	{
 		ID: "2019-10-09-project-description",
 		Migrate: func(tx *gorm.DB) error {
-			return tx.AutoMigrate(&models.Project{}).Error
+			return tx.Migrator().DropTable(&models.Project{})
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/repositories/config/migrations.go]: Thanks for catching this. Copy paste issue :D

File: cmd/entrypoints/clusterresource.go [hunk 9]
```diff
@@ -42,7 +42,8 @@ var controllerRunCmd = &cobra.Command{
 		dbConfigValues := configuration.ApplicationConfiguration().GetDbConfig()
 		dbConfig := repositoryConfig.DbConfig{
 			BaseConfig: repositoryConfig.BaseConfig{
-				IsDebug: dbConfigValues.Debug,
+				LogLevel:                                 dbConfigValues.LogLevel,
+				DisableForeignKeyConstraintWhenMigrating: dbConfigValues.DisableForeignKeyConstraintWhenMigrating,
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply cmd/entrypoints/clusterresource.go]: DisableForeignKeyConstraintWhenMigrating is only required during migration. 
But i have kept the loglevel config which we are exposing through Debug flag in dbConfig

File: cmd/entrypoints/migrate.go [hunk 10]
```diff
@@ -44,16 +45,28 @@ var migrateCmd = &cobra.Command{
...
-		db.LogMode(true)
-		if err = db.DB().Ping(); err != nil {
+
+		sqlDB, err := db.DB()
+		if err != nil {
 			logger.Fatal(ctx, err)
 		}
 
+		defer func(deferCtx context.Context) {
+			if err = sqlDB.Close(); err != nil {
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply cmd/entrypoints/migrate.go:66]: I have kept this change to use Close() and we will have another issue to address the connection pooling feature of gorm. Didn't want to make too many changes on this PR . Let me know what you think

**Comment by pmahindrakar-oss** [comment]
> hey @pmahindrakar-oss thanks so much for this massive undertaking!
> 
> are you able to stand up this admin instance on a sandbox and verify that you can launch executions and use the console to navigate esp. with dynamic tasks, child workflows and other more complicated cases?

Hi @katrogan , I have been testing these changes using flytectl and flyteconsole.
Usual steps of registration, execution creation and fetching the execution were done.
Viewing these flyteconsole was also done.

The type of workflows executed were simple go_greet task and control flow core.control_flow.subworkflows.my_subwf.
Let me know specific workflows from cookbook which you think are needed in addition to these and whats covered in integration test pipeline.

File: pkg/repositories/config/seed_data.go [hunk 11]
```diff
@@ -18,7 +18,7 @@ func SeedProjects(db *gorm.DB, projects []string) error {
 			Name:        project,
 			Description: fmt.Sprintf("%s description", project),
 		}
-		if err := tx.Where(models.Project{Identifier: project}).FirstOrCreate(&projectModel).Error; err != nil {
+		if err := tx.Where(models.Project{Identifier: project}).Omit("id").FirstOrCreate(&projectModel).Error; err != nil {
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/repositories/config/seed_data.go:21]: The reason being , gorm is not using the sequence when inserting the autoIncrement fields. I tried multiple annotations old as well new . Through they mention autoIncrement in the docs https://gorm.io/docs/models.html 
I couldn't get it to work though hence relying on the sequence created for the ID column. This is done across the board for all tables. 


The way this was manifesting was all records were inserted with 0 ID and this was caught way later during integration tests of Parent-Child workflow. None of the unit tests actually caught this issue

File: pkg/repositories/config/migrations.go [hunk 12]
```diff
@@ -228,55 +237,49 @@ var Migrations = []*gormigrate.Migration{
...
 		ID: "2020-07-31-node-execution",
 		Migrate: func(tx *gorm.DB) error {
-			return tx.AutoMigrate(&models.NodeExecution{}).Error
-		},
-		Rollback: func(tx *gorm.DB) error {
-			return tx.Model(&models.NodeExecution{}).DropColumn("parent_id").DropColumn("node_execution_metadata").Error
-		},
-	},
-	{
-		ID: "2020-08-13-node-execution",
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/repositories/config/migrations.go:247]: These two from the migration script look exactly the same.
Screenshot 2021-12-15 at 5 23 51 PM

May be just an error while commiting ?

File: pkg/repositories/config/postgres.go [hunk 13]
```diff
@@ -57,26 +56,23 @@ func (p *PostgresConfigProvider) GetArgs() string {
...
+	db, err := gorm.Open(postgres.Open(config.GetDSN()), &gorm.Config{
+		Logger:                                   logger.Default.LogMode(config.GetDBConfig().LogLevel),
+		DisableForeignKeyConstraintWhenMigrating: config.GetDBConfig().DisableForeignKeyConstraintWhenMigrating,
+	})
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
-	db.LogMode(config.IsDebug())
-	validations.RegisterCallbacks(db)
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/repositories/config/postgres.go:80]: Its no longer supported.
the library used is using v1 gorm https://github.com/qor/validations/blob/master/callbacks.go
Dont find any updates to this.  I can create a separate issue to add support v2. Let me know if you know of any alternatives

File: pkg/repositories/models/node_execution.go [hunk 14]
```diff
@@ -28,19 +28,18 @@ type NodeExecution struct {
...
 	NodeExecutionMetadata []byte
 	// Parent that spawned this node execution - value is empty for executions at level 0
 	ParentID *uint `sql:"default:null" gorm:"index"`
 	// List of child node executions - for cases like Dynamic task, sub workflow, etc
-	ChildNodeExecutions []NodeExecution `gorm:"foreignkey:ParentID"`
+	ChildNodeExecutions []NodeExecution `gorm:"foreignKey:ParentID;references:ID"`
 	// The task execution (if any) which launched this node execution.
 	// TO BE DEPRECATED - as we have now introduced ParentID
-	ParentTaskExecutionID uint `sql:"default:null" gorm:"index"`
+	ParentTaskExecutionID *uint `sql:"default:null" gorm:"index"`
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/repositories/models/node_execution.go:39]: Here we are using this column as pointer
https://github.com/flyteorg/flyteadmin/blob/f60baa3c6c56b17c67d8c492188d56b390899d46/pkg/manager/impl/node_execution_manager.go#L74

whereas when its saved in DB it was saved as 0 with uint and not as null
This was causing the records to not be fetched when using this column as filter i.e isParent filter.

Changing this to pointer fixed it.

This is inline with how shared.ParentID:              nil,
is used which is used as pointer and correctly saves as null value if its not set.

I think earlier version of gorm would save it as null or allow filtering with nil as value 0 and hence it was working fine in v1

File: cmd/entrypoints/migrate.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (katrogan) [reply cmd/entrypoints/migrate.go:66]: got it, sounds good!

File: pkg/repositories/models/node_execution.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/repositories/models/node_execution.go:39]: 👍 thanks for explaining

File: pkg/repositories/config/migrations.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/repositories/config/migrations.go:247]: oh yeah good call. thank you for checking

File: pkg/repositories/config/postgres.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/repositories/config/postgres.go:80]: oh so the validations are now no longer being enforced? yes can we file an issue to re-add this functionality in a separate PR?

File: pkg/repositories/config/seed_data.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/repositories/config/seed_data.go:21]: >  hence relying on the sequence created for the ID column

can you elaborate what you mean by this? so auto increment is still working?

File: pkg/repositories/models/execution.go (hunk 3 above)
Comment (katrogan) [reply pkg/repositories/models/execution.go:37]: ah cool i think you're right

**Comment by katrogan** [comment]
thank you for all the testing @pmahindrakar-oss! would you mind running a few more to verify other relationships
- launch plans called in a workflow: https://github.com/flyteorg/flytesnacks/blob/dynamic-lp-subwf/cookbook/core/control_flow/nested_dynamic.py#L21
- parent workflow for my_subwf: https://github.com/flyteorg/flytesnacks/blob/561b49c8c2769b7d905f3c15717209708f771161/cookbook/core/control_flow/subworkflows.py#L77
and also try recovering and/or relaunching and ensuring the UI points back to the original workflow?

File: pkg/repositories/config/seed_data.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/repositories/config/seed_data.go:21]: > thank you for all the testing @pmahindrakar-oss! would you mind running a few more to verify other relationships
> 
> * launch plans called in a workflow: https://github.com/flyteorg/flytesnacks/blob/dynamic-lp-subwf/cookbook/core/control_flow/nested_dynamic.py#L21
> * parent workflow for my_subwf: https://github.com/flyteorg/flytesnacks/blob/561b49c8c2769b7d905f3c15717209708f771161/cookbook/core/control_flow/subworkflows.py#L77
>   and also try recovering and/or relaunching and ensuring the UI points back to the original workflow?

Here are the results

### launch plans called in a workflow
Screenshot 2021-12-16 at 5 27 58 PM

### launch plans called in a workflow - relaunch
Screenshot 2021-12-16 at 5 30 40 PM

### launch plans called in a workflow - aborted
Screenshot 2021-12-16 at 5 31 49 PM

### launch plans called in a workflow - recovered
Screenshot 2021-12-16 at 5 35 43 PM

### parent workflow for my_subwf

Screenshot 2021-12-16 at 5 40 01 PM

### parent workflow for my_subwf - relaunch
Screenshot 2021-12-16 at 5 47 51 PM


### parent workflow for my_subwf - aborted
Screenshot 2021-12-16 at 5 49 07 PM


### parent workflow for my_subwf - recoverd
 Screenshot 2021-12-16 at 5 54 58 PM

File: pkg/repositories/errors/postgres.go [hunk 15]
```diff
@@ -45,33 +48,40 @@ type postgresErrorTransformer struct {
...
-func (p *postgresErrorTransformer) ToFlyteAdminError(err error) errors.FlyteAdminError {
-	pqError, ok := err.(*pq.Error)
+func (p *postgresErrorTransformer) ToFlyteAdminError(err error) flyteAdminErrors.FlyteAdminError {
+	if unwrappedErr := errors.Unwrap(err); unwrappedErr != nil {
+		err = unwrappedErr
+	}
+
+	pqError, ok := err.(*pgconn.PgError)
 	if !ok {
+		logger.InfofNoCtx("Unable to cast to pgconn.PgError. Error type: [%v]",
```
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/repositories/errors/postgres.go]: Done

File: pkg/repositories/config/postgres.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/repositories/config/postgres.go:80]: https://github.com/flyteorg/flyte/issues/1955

File: cmd/entrypoints/migrate.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply cmd/entrypoints/migrate.go:66]: https://github.com/flyteorg/flyte/issues/1956

File: pkg/repositories/gormimpl/execution_repo.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply pkg/repositories/gormimpl/execution_repo.go:47]: Created an issue for it https://github.com/flyteorg/flyte/issues/1958
And will take it as a followup

**Comment by pmahindrakar-oss** [comment]
> thank you for all the testing @pmahindrakar-oss! would you mind running a few more to verify other relationships
> 
> * launch plans called in a workflow: https://github.com/flyteorg/flytesnacks/blob/dynamic-lp-subwf/cookbook/core/control_flow/nested_dynamic.py#L21
> * parent workflow for my_subwf: https://github.com/flyteorg/flytesnacks/blob/561b49c8c2769b7d905f3c15717209708f771161/cookbook/core/control_flow/subworkflows.py#L77
>   and also try recovering and/or relaunching and ensuring the UI points back to the original workflow?

Here are the results @katrogan 

### launch plans called in a workflow
Screenshot 2021-12-16 at 5 27 58 PM

### launch plans called in a workflow  - relaunch
Screenshot 2021-12-16 at 5 30 40 PM


### launch plans called in a workflow - aborted
Screenshot 2021-12-16 at 5 31 49 PM


### launch plans called in a workflow - recovered

Screenshot 2021-12-16 at 5 35 43 PM

### parent workflow for my_subwf
Screenshot 2021-12-16 at 5 40 01 PM

### parent workflow for my_subwf - relaunch
Screenshot 2021-12-16 at 5 47 51 PM

### parent workflow for my_subwf - aborted
Screenshot 2021-12-16 at 5 49 07 PM
 

### parent workflow for my_subwf - recovered
Screenshot 2021-12-16 at 5 54 58 PM

**Review by katrogan — APPROVED** [review]
